src/Module.Server/Rewards/CrpgRewardServer.cs

```diff
@@ -195,17 +210,78 @@ public override void OnRemoveBehavior()
             return;
         }
 
+        CalculatePlayerUpkeepCompensations();
+        CompensateTeamhitPlayers(userUpdates);
+
         // TODO: add retry mechanism (the endpoint need to be idempotent though).
         try
         {
-            var res = (await _crpgClient.UpdateUsersAsync(new CrpgGameUsersUpdateRequest { Updates = userUpdates })).Data!;
+            var res = (await _crpgClient.UpdateUsersAsync(new CrpgGameUsersUpdateRequest { Updates = userUpdates.Values.ToList() })).Data!;
             SendRewardToPeers(res.UpdateResults, crpgPeerByUserId, valorousPlayerIds);
         }
         catch (Exception e)
         {
             Debug.Print("Couldn't update users: " + e);
             SendErrorToPeers(crpgPeerByUserId);
         }
+
+        _compensationTable.Clear();
+    }
+
+    private void CalculatePlayerUpkeepCompensations()
+    {
+        foreach (var affectedEntry in _agentsThatGotTeamhitByCrpgPeer)
+        {
+            Agent affectedAgent = affectedEntry.Value;
+            CrpgPeer affectedCrpgPeer = affectedEntry.Key;
+            foreach (Agent.Hitter affector in affectedAgent.HitterList)
+            {
+                CrpgPeer affectorCrpgPeer = affector.HitterPeer.GetNetworkPeer().GetComponent<CrpgPeer>();
+                if (affectedCrpgPeer == affectorCrpgPeer)
+                {
+                    continue;
+                }

```

Comment:
What happen if you get teamhit by a bot?

Reply:
The Method TryGetCrpgUserIdSafely will probably filter out bots, because i'm certain that those don't have a MissionPeer or NetworkPeer